csrc/ir_nodes.cpp

```diff
@@ -1329,6 +1329,177 @@ Val* GroupedWelfordOp::getInitValOfOutput(Val* output_val) const {
 
 NVFUSER_DEFINE_CLONE_AND_CREATE(GroupedWelfordOp)
 
+namespace MmaOpUtils {
+// A helper structure, used to populate details about inputs and output
+//  of MmaOp.
+struct MmaDetails {
+  constexpr static size_t UNINITIALIZED = 99;
+  size_t in_a_bcast = UNINITIALIZED;
+  size_t in_b_bcast = UNINITIALIZED;
+  size_t output_rdomain = UNINITIALIZED;
+  MmaOptions::MmaInputLayout layout = MmaOptions::MmaInputLayout::TT;
+};
+
+MmaDetails getMmaDetails(TensorView* out, TensorView* in_a, TensorView* in_b) {
+  MmaDetails details;
+
+  // NOTE: input tv has broadcast, checked in MmaOp::MmaOp
+  const auto findBcastInMmaInput = [](const TensorView* tv) {
+    size_t bcast = MmaDetails::UNINITIALIZED;
+    for (size_t pos = 0; pos < tv->nDims(); ++pos) {
+      if (tv->axis(static_cast<int>(pos))->isBroadcast()) {
+        if (bcast != MmaDetails::UNINITIALIZED) {
+          TORCH_INTERNAL_ASSERT(
+              false,
+              "One of MmaOp inputs has more than one broadcast iterdomain");
+        } else {
+          bcast = pos;
+        }
+      }
+    }
+    return bcast;
+  };
+  // NOTE: output tv has reduction, checked in MmaOp::MmaOp
+  const auto findRdomainInMmaOutput = [](const TensorView* tv) {
+    size_t rdomain = MmaDetails::UNINITIALIZED;
+    for (size_t pos = 0; pos < tv->nDims(); ++pos) {
+      if (tv->axis(static_cast<int>(pos))->isReduction()) {
+        if (rdomain != MmaDetails::UNINITIALIZED) {
+          TORCH_INTERNAL_ASSERT(
+              false, "MmaOp output has more than one reduction iterdomain");
+        } else {
+          rdomain = pos;
+        }
+      }
+    }
+    return rdomain;
+  };
+
+  details.in_a_bcast = findBcastInMmaInput(in_a);
+  details.in_b_bcast = findBcastInMmaInput(in_b);
+  details.output_rdomain = findRdomainInMmaOutput(out);
+
+  if (details.in_a_bcast == static_cast<size_t>(2) &&
+      details.in_b_bcast == static_cast<size_t>(0)) {
+    details.layout = MmaOptions::MmaInputLayout::TT;
+  } else if (
+      details.in_a_bcast == static_cast<size_t>(1) &&
+      details.in_b_bcast == static_cast<size_t>(0)) {
+    details.layout = MmaOptions::MmaInputLayout::TN;
+  } else if (
+      details.in_a_bcast == static_cast<size_t>(2) &&
+      details.in_b_bcast == static_cast<size_t>(1)) {
+    details.layout = MmaOptions::MmaInputLayout::NT;
+  } else {
+    TORCH_INTERNAL_ASSERT(false, "Unsupported input layout");
+  }
+
+  return details;
+}
+
+void validateMmaOpDefinition(
```

Comment:
Could you relax this check to be more tolerant to batched matmul, tensor contractions, and batched tensor contraction?

This is what I think we should do:

I think we should check `a`, `b`, and `out` to have the same number of dimensions, and classify these dimensions into four types:
1. What are the dimensions that is a broadcast in `b`, concrete in `a`, and not reduced in `out`?
2. What are the dimensions that is a broadcast in `a`, concrete in `b`, and not reduced in `out`?
3. What are the dimensions that is reduced in `out`?
4. Other dimensions.

The answer of `1` will be `M` dimensions, the answer of `2` will be `N` dimensions, and the answer of `3` will be `K` dimensions. And the dimensions not in `M`, `N` or `K` should be batch dimensions. In the ctor of `MMAOp`, I think we should allow `M`, `N` and `K` to be arbitrary size.

But we should have a method `std::vector<int64_t> getMAxes()` and similarly `getNAxes`, `getKAxes`, `getBatchAxes`, and in the matmul scheduler, we should check that `getMAxes().size() == 1 && getNAxes().size() == 1 && getKAxes().size() == 1 && getBatchAxes() == 0`.

What do you think?

Reply:
I like your idea of handling the problem - I need to check how to implement this in a efficient way. It won't be good to do the heavy processing on demand, we should calculate `M`/`N`/`K`/`Batches` in constructor and store it as private member structure. Data can be access through API you mentioned.